File: hackathon_CPP_SDK/src-gen/nevonex/gps_tc/GPS_TC.cpp
/*
 * nevonex/gps_tc/GPS_TC.cpp
 * Copyright (c) Robert Bosch GmbH. All rights reserved.
 */

#include "GPS_TC.hpp"
#include <ecore/EObject.hpp>
#include <ecore/EClass.hpp>
#include "nevonex/gps_tc/Gps_tcPackage.hpp"
#include <ecorecpp/mapping.hpp>
#include <nevonex-fcal-platform/log/Logger.hpp>

/*PROTECTED REGION ID(GPS_TC.cpp) START*/
// Please, enable the protected region if you add manually written code.
// To do this, add the keyword ENABLED before START.
/*PROTECTED REGION END*/

using namespace ::nevonex::gps_tc;
using namespace ::nevonex::log;

// Default constructor
GPS_TC::GPS_TC() : m_active_TC_GPS_source(0), m_positionofGpsSensor(0), m_tcGpsInfo(0)
{
    // Machine does NOT call _initialize() — Provider manages initialization
    m_interfaceDetailMap = {
    };

    /*PROTECTED REGION ID(GPS_TC__GPS_TC) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

GPS_TC::~GPS_TC()
{
    /*PROTECTED REGION ID(GPS_TC__Destructor) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

// References
::ecore::EString GPS_TC::getActive_TC_GPS_source() const
{
    return m_active_TC_GPS_source;
}

void GPS_TC::setActive_TC_GPS_source(::ecore::EString _active_TC_GPS_source)
{
    ::ecore::EString _old_active_TC_GPS_source = m_active_TC_GPS_source;
    m_active_TC_GPS_source = _active_TC_GPS_source;
    SDK_FCAL_LOG(SeverityLevel::debug) << "GPS_TC property \"active_TC_GPS_source\" has changed.";
    notifyPropertyChange("active_TC_GPS_source", _old_active_TC_GPS_source, m_active_TC_GPS_source);
}

::ecore::EString GPS_TC::getPositionofGpsSensor() const
{
    return m_positionofGpsSensor;
}

void GPS_TC::setPositionofGpsSensor(::ecore::EString _positionofGpsSensor)
{
    ::ecore::EString _old_positionofGpsSensor = m_positionofGpsSensor;
    m_positionofGpsSensor = _positionofGpsSensor;
    SDK_FCAL_LOG(SeverityLevel::debug) << "GPS_TC property \"positionofGpsSensor\" has changed.";
    notifyPropertyChange("positionofGpsSensor", _old_positionofGpsSensor, m_positionofGpsSensor);
}

::nevonex::fcal::GPS_INFO_R_ptr GPS_TC::getTcGpsInfo() const
{
    return m_tcGpsInfo;
}

void GPS_TC::setTcGpsInfo(::nevonex::fcal::GPS_INFO_R_ptr _tcGpsInfo)
{
    ::nevonex::fcal::GPS_INFO_R_ptr _old_tcGpsInfo = m_tcGpsInfo;
    m_tcGpsInfo = _tcGpsInfo;
    SDK_FCAL_LOG(SeverityLevel::debug) << "GPS_TC property \"tcGpsInfo\" has changed.";
    notifyPropertyChange("tcGpsInfo", _old_tcGpsInfo, m_tcGpsInfo);
}

// FCAL timestamp/valid
::ecore::ELong GPS_TC::getActive_TC_GPS_source_Timestamp()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).getTimestamp();
    return 0;
}

::ecore::EBoolean GPS_TC::isActive_TC_GPS_source_Valid()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).isValid();
    return false;
}

::ecore::ELong GPS_TC::getPositionofGpsSensor_Timestamp()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).getTimestamp();
    return 0;
}

::ecore::EBoolean GPS_TC::isPositionofGpsSensor_Valid()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).isValid();
    return false;
}

::ecore::ELong GPS_TC::getTcGpsInfo_Timestamp()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).getTimestamp();
    return 0;
}

::ecore::EBoolean GPS_TC::isTcGpsInfo_Valid()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).isValid();
    return false;
}

// PropertyChange delegation (diamond inheritance requires explicit override)
void GPS_TC::addPropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::addPropertyChangeListener(_listener);
}

void GPS_TC::removePropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::removePropertyChangeListener(_listener);
}

void GPS_TC::notifyPropertyChange(::ecore::EString const &_name,
        ::ecore::EJavaObject const &_oldValue,
        ::ecore::EJavaObject const &_newValue)
{
    using namespace ::nevonex::types;
    PropertyChange::notifyPropertyChange(_name, _oldValue, _newValue);
}

File: com.bosch.fsp.hackathon.gen/src/com/bosch/nevonex/gps_tc/impl/GPS_TC.java
/**
Copyright (c) Robert Bosch GmbH. All rights reserved.
*/
package com.bosch.nevonex.gps_tc.impl;

import com.bosch.fsp.logger.FCALLogs;
import com.bosch.fsp.logger.LoggerConstants;

import com.bosch.fsp.platform.InterfaceDetails;

import com.bosch.fsp.runtime.feature.IMachine;

import com.bosch.nevonex.common.ITopicObject;
import com.bosch.nevonex.common.impl.TopicObject;

import com.bosch.nevonex.fcal.IGPS_INFO_R;

import com.bosch.nevonex.gps_tc.IGPS_TC;

import com.bosch.nevonex.types.IPropertyChange;
import com.bosch.nevonex.types.impl.TypesPackage;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import java.lang.reflect.InvocationTargetException;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.eclipse.emf.common.util.BasicEList;
import org.eclipse.emf.common.util.EList;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.util.EDataTypeUniqueEList;

/**
 * An implementation of the model object '<em><b>GPS_TC</b></em>'.
 * @generated
 */
public class GPS_TC extends TopicObject implements IGPS_TC {
	protected EList<PropertyChangeListener> listeners;
	protected Map<String, InterfaceDetails> interfaceDetailMap = new HashMap<>();
	protected String active_TC_GPS_source = null;
	protected String positionofGpsSensor = null;
	protected IGPS_INFO_R tcGpsInfo;

	/**
	 * @generated
	 */
	protected GPS_TC() {
		super();
		initInterfaceDetails();
	}

	private void initInterfaceDetails() {
		// Interface details initialization
	}

	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return Gps_tcPackage.Literals.GPS_TC;
	}

	public String getActive_TC_GPS_source() {
		return active_TC_GPS_source;
	}

	public void setActive_TC_GPS_source(String newActive_TC_GPS_source) {
		active_TC_GPS_source = newActive_TC_GPS_source;
	}

	public String getPositionofGpsSensor() {
		return positionofGpsSensor;
	}

	public void setPositionofGpsSensor(String newPositionofGpsSensor) {
		positionofGpsSensor = newPositionofGpsSensor;
	}

	/**
	 * @generated
	 */
	public IGPS_INFO_R getTcGpsInfo() {
		if (tcGpsInfo != null && tcGpsInfo.eIsProxy()) {
			InternalEObject oldTcGpsInfo = (InternalEObject) tcGpsInfo;
			tcGpsInfo = (IGPS_INFO_R) eResolveProxy(oldTcGpsInfo);
			if (tcGpsInfo != oldTcGpsInfo) {
			}
		}
		return tcGpsInfo;
	}

	/**
	 * @generated
	 */
	public IGPS_INFO_R basicGetTcGpsInfo() {
		return tcGpsInfo;
	}

	/**
	 * @generated
	 */
	public void setTcGpsInfo(IGPS_INFO_R newTcGpsInfo) {
		IGPS_INFO_R oldTcGpsInfo = tcGpsInfo;
		tcGpsInfo = newTcGpsInfo;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " GPS_TC property \"tcGpsInfo\" changed from  "
						+ oldTcGpsInfo + " to " + newTcGpsInfo);
		notifyPropertyChange("tcGpsInfo", oldTcGpsInfo, newTcGpsInfo);
	}

	/**
	 * @generated
	 */
	public long getTcGpsInfo_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("tcGpsInfo");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isTcGpsInfo_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("tcGpsInfo");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public void addPropertyChangeListener(PropertyChangeListener listener) {
		if (listeners == null) {
			listeners = new BasicEList<>();
		}
		listeners.add(listener);
	}

	/**
	 * @generated
	 */
	public void removePropertyChangeListener(PropertyChangeListener listener) {
		if (listeners != null) {
			listeners.remove(listener);
		}
	}

	/**
	 * @generated
	 */
	public void notifyPropertyChange(String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(this, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
		case Gps_tcPackage.IGPS_TC__ACTIVE_TC_GPS_SOURCE:
			return getActive_TC_GPS_source();
		case Gps_tcPackage.IGPS_TC__POSITIONOF_GPS_SENSOR:
			return getPositionofGpsSensor();
		case Gps_tcPackage.IGPS_TC__TC_GPS_INFO:
			if (resolve)
				return getTcGpsInfo();
			return basicGetTcGpsInfo();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * @generated
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
		case Gps_tcPackage.IGPS_TC__ACTIVE_TC_GPS_SOURCE:
			setActive_TC_GPS_source((String) newValue);
			return;
		case Gps_tcPackage.IGPS_TC__POSITIONOF_GPS_SENSOR:
			setPositionofGpsSensor((String) newValue);
			return;
		case Gps_tcPackage.IGPS_TC__TC_GPS_INFO:
			setTcGpsInfo((IGPS_INFO_R) newValue);
			return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
		case Gps_tcPackage.IGPS_TC__ACTIVE_TC_GPS_SOURCE:
			setActive_TC_GPS_source(null);
			return;
		case Gps_tcPackage.IGPS_TC__POSITIONOF_GPS_SENSOR:
			setPositionofGpsSensor(null);
			return;
		case Gps_tcPackage.IGPS_TC__TC_GPS_INFO:
			setTcGpsInfo((IGPS_INFO_R) null);
			return;
		}
		super.eUnset(featureID);
	}

	/**
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
		case Gps_tcPackage.IGPS_TC__ACTIVE_TC_GPS_SOURCE:
			return active_TC_GPS_source != null;
		case Gps_tcPackage.IGPS_TC__POSITIONOF_GPS_SENSOR:
			return positionofGpsSensor != null;
		case Gps_tcPackage.IGPS_TC__TC_GPS_INFO:
			return tcGpsInfo != null;
		}
		return super.eIsSet(featureID);
	}
} //GPS_TC
